### PR TITLE
feat/cli-tab-completion: add bash-completion using tabtab package

### PR DESCRIPTION
**Pull request by spirAde**
This makes it possible to use bash-completion using tabtab package. At the moment, all possible options are included for all existing commands, except for general ones (such as --help, --force installation, etc). To activate bash-completion, a user needs to run the `lb4 install-completion` command, to remove the completion `lb4 uninstall-completion`.

Issue #2358

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Comment by spirAde** [comment]
Hello, guys. I would like to clarify the point with common commands (--help, --force-installation, etc) whether we should use them in bash-completion?

**Comment by spirAde** [comment]
Also, the user will need to restart the .bashrc, .zshrc after lb4 install-completion command, so that bash-completion began to work. What is the best way to inform the user about this and is it worth it?

File: packages/cli/lib/tab-completion.js [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+const debug = require('./debug')();
+const tabtab = require('tabtab');
+
+const tabCompletionCommands = [
+  'install-completion',
+  'uninstall-completion',
+  'completion',
+];
+
+const commands = [
```
Comment (raymondfeng) [inline packages/cli/lib/tab-completion.js:17]: I think it's possible to build these lists from generator metadata without hard-coding them. Let's look at how `help` is implemented. `--help` lists available commands and their options.

**Comment by raymondfeng** [comment]
@spirAde Please sign the CLA - https://cla.strongloop.com/agreements/strongloop/loopback-next. Thanks.

**Comment by spirAde** [comment]
> @spirAde Please sign the CLA - https://cla.strongloop.com/agreements/strongloop/loopback-next. Thanks.

I've already signed it
image

image

File: packages/cli/lib/tab-completion.js (hunk 1 above)
Comment (spirAde) [reply packages/cli/lib/tab-completion.js:17]: @raymondfeng Thank you for your feedback. I've checked the current solution for getting all possible commands. And I was thinking about it before to take `env.getGeneratorsMeta()` solution. But I see a problem with this solution. Even if I get commands from `env.getGeneratorsMeta()`, we don't have the possibility to take command's options from yeoman. Therefore hardcoded options and not-hardcoded commands should have been supported that is not good from my perspective.

But I investigated `yeoman-environment` and found internal `_options` property. What do you think about this proposal: add `getOptions` method inside `BaseGenerator` class.

```
...
getOptions() {
  return this._options;
}
...
```

this._options view for controller:

```
{ help:
   { name: 'help',
     type: [Function: Boolean],
     alias: 'h',
     description: 'Print the generator\'s options and usage' },
  'skip-cache':
   { name: 'skip-cache',
     type: [Function: Boolean],
     description: 'Do not remember prompt answers',
     default: false },
  'skip-install':
   { name: 'skip-install',
     type: [Function: Boolean],
     description: 'Do not automatically install dependencies',
     default: false },
  'force-install':
   { name: 'force-install',
     type: [Function: Boolean],
     description: 'Fail on install dependencies error',
     default: false },
  'ask-answered':
   { type: [Function: Boolean],
     description: 'Show prompts for already configured options',
     default: false,
     name: 'ask-answered',
     hide: false },
  controllerType:
   { type: [Function: String],
     required: false,
     description: 'Type for the controller',
     name: 'controllerType',
     hide: false },
  config:
   { type: [Function: String],
     alias: 'c',
     description: 'JSON file name or value to configure options',
     name: 'config',
     hide: false },
  yes:
   { type: [Function: Boolean],
     alias: 'y',
     description:
      'Skip all confirmation prompts with default or provided value',
     name: 'yes',
     hide: false },
  format:
   { type: [Function: Boolean],
     description: 'Format generated code using npm run lint:fix',
     name: 'format',
     hide: false } }

```

After getting options, I can filter and convert them to a fit view. It makes possible to use dynamic commands and options.

Comment (raymondfeng) [reply packages/cli/lib/tab-completion.js:17]: I'm adding code to get options and args for each command.

**Comment by spirAde** [comment]
> @spirAde Please sign the CLA - https://cla.strongloop.com/agreements/strongloop/loopback-next. Thanks.

I've already signed. 
image
image
image

**Comment by raymondfeng** [comment]
@spirAde I rebased the code against latest master and pushed one commit to use inspected options and arguments. Please continue from that. Please also add some tests.

**Comment by raymondfeng** [comment]
You may need to run the following git commands to pull it down:
- git fetch --all
- git reset --hard origin/feat/cli-tab-completion

File: packages/cli/lib/tab-completion.js (hunk 1 above)
Comment (spirAde) [reply packages/cli/lib/tab-completion.js]: It's better solution) Thank you

**Comment by raymondfeng** [comment]
@spirAde The code can be further simplified as follows:

```
/**
 * Get all available commands or options depends on entered input
 * @param partial - completion env.partial
 */
function generateCompletions(optionsAndArgs, partial) {
  const commands = optionsAndArgs.map(i => i.name);
  const enteredCommand = commands.find(command => partial.includes(command));

  if (!enteredCommand) {
    return commands;
  }

  const options = optionsAndArgs[enteredCommand].map(o => o.name);
  const enteredOptions = options.filter(option => partial.includes(option));

  if (!enteredOptions.length) {
    return optionsAndArgs[enteredCommand].options.map(o => `--${o.name}`);
  }

  return options.filter(option => !enteredOptions.includes(option.name));
}

I tried and it seems that getting options take some time. The responsiveness is good. We need to cache the options to the file system.
```

**Comment by spirAde** [comment]
> @spirAde The code can be further simplified as follows:
> 
> ```js
> /**
>  * Get all available commands or options depends on entered input
>  * @param partial - completion env.partial
>  */
> function generateCompletions(optionsAndArgs, partial) {
>   const commands = optionsAndArgs.map(i => i.name);
>   const enteredCommand = commands.find(command => partial.includes(command));
> 
>   if (!enteredCommand) {
>     return commands;
>   }
> 
>   const options = optionsAndArgs[enteredCommand].map(o => o.name);
>   const enteredOptions = options.filter(option => partial.includes(option));
> 
>   if (!enteredOptions.length) {
>     return optionsAndArgs[enteredCommand].options.map(o => `--${o.name}`);
>   }
> 
>   return options.filter(option => !enteredOptions.includes(option.name));
> }
> ```
> 
> I tried and it seems that getting options takes some time. The responsiveness is NOT good. We need to cache the options to the file system.

Yes, that was another reason why I preferred hardcoded data. Time measurement gave out ~ 200ms.

**Comment by spirAde** [comment]
I will write when I get some results

**Comment by raymondfeng** [comment]
one idea is to save the metadata into a json file at release time so that we don’t have to rebuild when tab is hit.

**Comment by raymondfeng** [comment]
@spirAde I went ahead to add one commit:

1. Allow `lb4 --meta` to dump command options to `.yo-rc.json`
2. `lb4 completion` uses information from `.yo-rc.json` for completion
3. I made `lb4 completion` as fast as possible without loading other things

**Comment by raymondfeng** [comment]
@spirAde The first commit was made under `dspiridonov4xxi <dspiridonov@4xxi.com>`. That's probably why the CLA complains. Can we amend the commit to use your github account?

**Comment by spirAde** [comment]
@raymondfeng Thank you for your help. I have measured responsiveness on my local machines(macOS, Linux) and can say it's more than enough. I added tests to reach 100% coverage for `lib/tab-completion`. But I don't understand why CI process fails. I'd checked all tests on my both machines and didn't get any errors.

**Comment by raymondfeng** [comment]
@spirAde Thank you for the update. Don't worry about the CI failures - they are caused by timeouts. I'll increase the timeout for CI.

I'll squash the commits and merge it.

Again, thank you for the great contribution! We love ideas to improve developer experience for LoopBack! 👍

**Pull request merged**